File: ios/Packages/CoreAuth/Sources/CoreAuth/BiometricGate.swift
import Foundation
import LocalAuthentication

/// Garde biométrique (FaceID / TouchID) pour les actions sensibles.
///
/// Utilisée avant un **retrait** ou un **changement de PIN wallet** (exigence sécurité).
/// La biométrie s'ajoute par-dessus le PIN de retrait 6 chiffres déjà vérifié côté backend.
public struct BiometricGate {

    public enum BiometricError: Error {
        case unavailable          // pas de capteur / non configuré
        case failed               // échec d'authentification
        case cancelled            // annulé par l'utilisateur
    }

    public init() {}

    /// Type de biométrie disponible (pour adapter le libellé/icone UI).
    public var availableType: LABiometryType {
        let context = LAContext()
        _ = context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: nil)
        return context.biometryType // .faceID / .touchID / .none
    }

    /// Demande une authentification biométrique.
    /// - Parameter reason: raison affichée à l'utilisateur (ex. « Confirmer le retrait »).
    /// - Throws: `BiometricError` selon l'issue.
    public func authenticate(reason: String) async throws {
        let context = LAContext()
        context.localizedFallbackTitle = "" // pas de repli code (biométrie stricte)

        var authError: NSError?
        guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &authError) else {
            throw BiometricError.unavailable
        }
        do {
            let ok = try await context.evaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, localizedReason: reason)
            if !ok { throw BiometricError.failed }
        } catch let laError as LAError where laError.code == .userCancel || laError.code == .appCancel {
            throw BiometricError.cancelled
        } catch {
            throw BiometricError.failed
        }
    }
}

File: android/core/auth/src/main/kotlin/com/dualmusic/core/auth/BiometricGate.kt
package com.dualmusic.core.auth

import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine

/**
 * Garde biométrique (empreinte / visage) pour les actions sensibles.
 *
 * Utilisée avant un **retrait** ou un **changement de PIN wallet**. S'ajoute par-dessus le
 * PIN de retrait 6 chiffres déjà vérifié côté backend.
 */
class BiometricGate(private val activity: FragmentActivity) {

    sealed class BiometricError(message: String) : Throwable(message) {
        object Unavailable : BiometricError("Biométrie indisponible")
        object Failed : BiometricError("Authentification échouée")
        object Cancelled : BiometricError("Annulé")
    }

    /** Vrai si une biométrie forte est configurée sur l'appareil. */
    fun isAvailable(): Boolean =
        BiometricManager.from(activity)
            .canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG) ==
            BiometricManager.BIOMETRIC_SUCCESS

    /**
     * Demande une authentification biométrique.
     * @param title titre du prompt (ex. « Confirmer le retrait »).
     * @param subtitle sous-titre optionnel.
     * @throws BiometricError selon l'issue.
     */
    suspend fun authenticate(title: String, subtitle: String? = null) {
        if (!isAvailable()) throw BiometricError.Unavailable

        suspendCancellableCoroutine { cont ->
            val executor = ContextCompat.getMainExecutor(activity)
            val prompt = BiometricPrompt(activity, executor, object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    if (cont.isActive) cont.resume(Unit)
                }
                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    if (!cont.isActive) return
                    val cancel = errorCode == BiometricPrompt.ERROR_USER_CANCELED ||
                        errorCode == BiometricPrompt.ERROR_NEGATIVE_BUTTON
                    cont.resumeWithException(if (cancel) BiometricError.Cancelled else BiometricError.Failed)
                }
                override fun onAuthenticationFailed() { /* tentative rejetée : le prompt reste ouvert */ }
            })

            val info = BiometricPrompt.PromptInfo.Builder()
                .setTitle(title)
                .apply { subtitle?.let { setSubtitle(it) } }
                .setNegativeButtonText("Annuler")
                .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
                .build()

            prompt.authenticate(info)
        }
    }
}
